### PR TITLE
Support manual service configuration

### LINKED ISSUE #23: Support custom authentication endpoints
We are currently relying on the openID [discovery endpoint](https://openid.net/specs/openid-connect-discovery-1_0.html) to infer the authorisation and revocation urls. Whereas this is very convenient most of the time, there can be cases where the discovery endpoint does not exist which makes it impossible to use this library.

It should therefore be possible to configure an auth session with custom url schemes rather than fetching from the discovery endpoint, which would make this library usable for services that are not openid-compliant.

### PR DESCRIPTION
**Pull request by jevakallio**
Fixes #23.

Adds support for manually configuring service endpoints and supporting client secrets for non-OIDC-compliant OAuth services. 

## JavaScript
- [x] Add serviceConfiguration to API
- [x] Add clientSecret to API
- [x] Support `authorize`
- [x] Support `refresh`
- [x] Support `revoke`
- [x] Refactor
- [x] Update and add tests

## Android
- [x] Allow authorization with manual service configuration
- [x] Allow refresh with manual service configuration
- [x] Pass additionalProperties to token request to support client_secret
- [x] Refactor

## iOS
- [x] Allow authorization with manual service configuration
- [x] Allow refresh with manual service configuration
- [x] Pass client_secret

## Other
- [x] Update API documentation
- [x] Update TypeScript definitions
- [x] Add note/warning about using client secrets
- [x] Add example configuration for a manually configured service

With this, we should next be able to implement: 
- #28 Identity Server 3 compatibility
- #42 More presets/examples

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -72,14 +72,20 @@ const result = await authorize(config);
 This is your configuration object for the client. The config is passed into each of the methods
 with optional overrides.
 
-* **issuer** - (`string`) _REQUIRED_ the url of the auth server
+* **issuer** - (`string`) base URI of the authentication server. If no `serviceConfiguration` (below) is not provided, issuer is a mandatory field, so that the configuration can be fetched from the issuer's [OIDC discovery endpoint](https://openid.net/specs/openid-connect-discovery-1_0.html).
```
Comment (kadikraman) [inline README.md]: "If no `serviceConfiguration` (below) is not provided" - double negative 🤔

File: README.md [hunk 2]
```diff
@@ -467,6 +481,42 @@ const refreshedState = await refresh(config, {
 });
 ```
 
+### Uber
+
+Uber provides an OAuth 2.0 endpoint for logging in with a Uber user's credentiala. You'll need to first [create an Uber OAuth application here](https://developer.uber.com/docs/riders/guides/authentication/introduction).
```
Comment (kadikraman) [inline README.md]: Uber's *credentials*

File: android/src/main/java/com/reactlibrary/RNAppAuthModule.java [hunk 3]
```diff
@@ -25,169 +24,101 @@
...
+                authorizeWithConfiguration(
+                        createAuthorizationServiceConfiguration(serviceConfiguration),
+                        appAuthConfiguration,
+                        clientId,
+                        scopesString,
+                        redirectUrl,
+                        additionalParametersMap
+                );
+            } catch (Exception e) {
+                promise.reject(e);
```
Comment (kadikraman) [inline android/src/main/java/com/reactlibrary/RNAppAuthModule.java]: In other parts of the app we've been more explicit with the error, e.g.
```
promise.reject("RNAppAuth Error", "Failed to authenticate", e);
```

File: android/src/main/java/com/reactlibrary/utils/MapUtils.java [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+
+import android.support.annotation.Nullable;
+
+import com.facebook.react.bridge.ReadableMap;
+import com.facebook.react.bridge.ReadableMapKeySetIterator;
+
+import java.util.HashMap;
+
+/**
+ * Created by formidable on 21/02/2018.
```
Comment (kadikraman) [inline android/src/main/java/com/reactlibrary/utils/MapUtils.java]: Not sure these "Created by" comments are particularly useful.

File: index.js [hunk 5]
```diff
@@ -5,8 +5,20 @@ const { RNAppAuth } = NativeModules;
...
   invariant(scopes && scopes.length, 'Scope error: please add at least one scope');
-const validateIssuer = issuer =>
-  invariant(typeof issuer === 'string', 'Config error: issuer must be a string');
+const validateIssuerOrServiceConfigurationEndpoints = (issuer, serviceConfiguration) =>
+  invariant(
+    typeof issuer === 'string' ||
+      (serviceConfiguration &&
+        typeof serviceConfiguration.authorizationEndpoint === 'string' &&
+        typeof serviceConfiguration.tokenEndpoint === 'string'),
+    'Config error: you must provide either an issue or a service endpoints'
```
Comment (kadikraman) [inline index.js]: "an issuer or a service endpoint"

File: index.js [hunk 6]
```diff
@@ -5,8 +5,20 @@ const { RNAppAuth } = NativeModules;
...
+  );
+const validateIssuerOrServiceConfigurationRevocationEndpoint = (issuer, serviceConfiguration) =>
+  invariant(
+    typeof issuer === 'string' ||
+      (serviceConfiguration && typeof serviceConfiguration.revocationEndpoint === 'string'),
+    'Config error: you must provide either an issue or a revocation endpoint'
```
Comment (kadikraman) [inline index.js]: issue__r__

**Comment by jevakallio** [comment]
@kadikraman fixed review notes, thanks! 

When you merge this, feel free to squash, the commit messages are not particularly useful.

**Pull request merged**